WeatherFlow Forecast (REST API)

jeeftor opened the pull request:
## Proposed change

<img width="584" alt="image" src="https://github.com/home-assistant/core/assets/6491743/51f04cc6-c62e-4f12-9322-fc8b62ab54c2">


I wrote the `weatherflow`  integration which is a **local** only integration that receives weather data via UDP. Tempest Weatherflow also provides a Cloud/REST API. Initially my plan was to port over the HACS module written by @briis - however - in realized my life would just be simpler to write a new backing library [weatherflow4py](https://github.com/jeeftor/weatherflow4py) 

Additionally - in discussions with @joostlek - we determined the best approach was to keep things as simple as possible - so following this path is how I ended up with a very simple backing library and config flow (we just store a single token).

### Hybrid vs Separate Integrations

I had a discussion with @joostlek about whether I should create a separate integration or extend the existing integration. I started down the path of extending the existing integration but the overall approach to these two integrations is different enough I think it serves to have it as a separate integration. One is entirely udp based while this one is entirely http based. I could also anticipate some users would want to run both at the same time



## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/30550

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

joostlek commented on homeassistant/components/weatherflow_cloud/config_flow.py:
```diff
@@ -0,0 +1,61 @@
+from .const import DOMAIN
+
+
+async def _async_validate_api_token(api_token) -> bool:
```
Can be inlined

joostlek commented on homeassistant/components/weatherflow_cloud/config_flow.py:
```diff
@@ -0,0 +1,61 @@
+
+    VERSION = 1
+
+    async def async_step_user(
```
Maybe flipping the logic in this step makes it more readable and easier to manage, (So from `if user_input is None:` to `if user_input is not None:`)

joostlek commented on homeassistant/components/weatherflow_cloud/config_flow.py:
```diff
@@ -0,0 +1,61 @@
+        if user_input is None:
+            return await self._show_setup_form(user_input)
+
+        await self.async_set_unique_id(user_input[CONF_API_TOKEN])
```
Can the api token change? Can you revoke it? If so, this isn't a good unique_id

jeeftor replied:
The token doesn't ever change - but it can be revoked.

- A token maps to a "set" of devices.
- If you add more devices to your account then they would be attached to existing tokens...
- There doesn't seem to be any other way to uniquely identify a set of devices... 
- A token can be deleted but they never expire

My thought process was I don't want to allow a user to enter the same token twice, but if say I had multipel tokens (my house, my rental, my parents) I'd want to allow for 3 entries into HA.

Any thoughts on alternatives?

gjohansson-ST replied:
Use `_async_abort_entries_match()` instead and make a reauth flow so the token can be fixed if neccessary.

joostlek replied:
Agreed

jeeftor replied:
Like so or do I nix the `self._abort_if_unique_id_configured()`


![image](https://github.com/home-assistant/core/assets/6491743/78efd421-d713-4d72-b482-8ac3b021c3af)

joostlek replied:
Remove both unique_id lines

joostlek replied:
And add pass user_input to the `self._async_abort_entries_match()`

joostlek commented on homeassistant/components/weatherflow_cloud/config_flow.py:
```diff
@@ -0,0 +1,61 @@
+        api_token = user_input[CONF_API_TOKEN]
+        try:
+            await _async_validate_api_token(api_token)
+            return self.async_create_entry(
```
Please only put stuff in the try block that can raise

joostlek commented on homeassistant/components/weatherflow_cloud/const.py:
```diff
@@ -0,0 +1,8 @@
+import logging
+
+DOMAIN = "weatherflow_cloud"
+_LOGGER = logging.getLogger(__package__)
```
module wide logger shouldn't be private
```suggestion
LOGGER = logging.getLogger(__package__)
```

joostlek commented on homeassistant/components/weatherflow_cloud/coordinator.py:
```diff
@@ -0,0 +1,46 @@
+from .const import _LOGGER, DOMAIN
+
+
+class WeatherFlowCloudDataUpdateCoordinator(DataUpdateCoordinator):
```
```suggestion
class WeatherFlowCloudDataUpdateCoordinator(DataUpdateCoordinator[dict[int, WeatherFlowData]]):
```

joostlek commented on homeassistant/components/weatherflow_cloud/coordinator.py:
```diff
@@ -0,0 +1,46 @@
+
+        self.data: dict[int, WeatherFlowData] = {}
+
+        super().__init__(
```
If you put this to the top of the file, `self.config_entry` will be set by the constructor of the DUC. Please overwrite the type in the coordinator, check the Youtube coordinator for an example

jeeftor replied:
I think I got this working...

joostlek commented on homeassistant/components/weatherflow_cloud/coordinator.py:
```diff
@@ -0,0 +1,46 @@
+        api_token = config_entry.data[CONF_API_TOKEN]
+        self.weather_api = WeatherFlowRestAPI(api_token=api_token)
+
+        self.data: dict[int, WeatherFlowData] = {}
```
```suggestion
```

joostlek commented on homeassistant/components/weatherflow_cloud/coordinator.py:
```diff
@@ -0,0 +1,46 @@
+        """Fetch data from WeatherFlow Forecast."""
+        try:
+            async with self.weather_api:
+                self.data = await self.weather_api.get_all_data()
+                return self.data
```
whatever you return in this function will be set to `self.data` by the superclass, so please only return here

joostlek commented on homeassistant/components/weatherflow_cloud/coordinator.py:
```diff
@@ -0,0 +1,46 @@
+            async with self.weather_api:
+                self.data = await self.weather_api.get_all_data()
+                return self.data
+        except Exception as err:
```
Please be more specific, catching bare exceptions is only allowed in the config flow

joostlek commented on homeassistant/components/weatherflow_cloud/manifest.json:
```diff
@@ -0,0 +1,13 @@
+  "name": "WeatherflowCloud",
+  "codeowners": ["@jeeftor"],
+  "config_flow": true,
+  "dependencies": [],
+  "documentation": "https://www.home-assistant.io/integrations/weatherflow_cloud",
+  "homekit": {},
+  "iot_class": "cloud_polling",
+  "requirements": ["weatherflow4py==0.1.10"],
+  "ssdp": [],
+  "zeroconf": []
```
Please remove empty fields

joostlek commented on homeassistant/components/weatherflow_cloud/weather.py:
```diff
@@ -0,0 +1,145 @@
+
+        self.station_id = station_id
+        self._attr_unique_id = f"weatherflow_forecast_{station_id}"
+        self._attr_name = self.local_data.station.name
```
On naming: I'd suggest naming the device to the station name. Then you can set `_attr_name = None`.

joostlek commented on homeassistant/components/weatherflow_cloud/weather.py:
```diff
@@ -0,0 +1,145 @@
+        self._attr_unique_id = f"weatherflow_forecast_{station_id}"
+        self._attr_name = self.local_data.station.name
+
+        outdoor_device = [
+            d for d in self.local_data.station.devices if d.device_type == "ST"
+        ][0]
```
What are we doing here?

jeeftor replied:
Getting the id from 1st outdoor sensor ... added a comment and a new method

jeeftor committed:
Update homeassistant/components/weatherflow_cloud/const.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

jeeftor committed:
Addressing 1st round of PR Comments

joostlek commented on homeassistant/components/weatherflow_cloud/config_flow.py:
```diff
@@ -0,0 +1,52 @@
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle a flow initialized by the user."""
+        if user_input is not None:
+            await self.async_set_unique_id(user_input[CONF_API_TOKEN])
+            self._abort_if_unique_id_configured()
+
+            # Validate Entry
+            api_token = user_input[CONF_API_TOKEN]
+            try:
+                async with WeatherFlowRestAPI(api_token) as api:
+                    await api.async_get_stations()
+            except ClientResponseError as err:
+                if err.status == 401:
+                    return await self._show_setup_form({"base": "invalid_api_key"})
+                return await self._show_setup_form({"base": "cannot_connect"})
+            return self.async_create_entry(
+                title="Weatherflow REST",
+                data={CONF_API_TOKEN: api_token},
+            )
+
+        return await self._show_setup_form(user_input)
+
+    async def _show_setup_form(self, errors=None):
+        """Show the setup form to the user."""
+        return self.async_show_form(
+            step_id="user",
+            data_schema=vol.Schema({vol.Required(CONF_API_TOKEN): str}),
+            errors=errors or {},
+        )
```
```suggestion
        errors = {}
        if user_input is not None:
            await self.async_set_unique_id(user_input[CONF_API_TOKEN])
            self._abort_if_unique_id_configured()

            # Validate Entry
            api_token = user_input[CONF_API_TOKEN]
            try:
                async with WeatherFlowRestAPI(api_token) as api:
                    await api.async_get_stations()
            except ClientResponseError as err:
                if err.status == 401:
                    errors["base"] = "invalid_api_key"
                else:
                    errors["base"] = "cannot_connect"
            else:
                return self.async_create_entry(
                    title="Weatherflow REST",
                    data={CONF_API_TOKEN: api_token},
                )

        return self.async_show_form(
            step_id="user",
            data_schema=vol.Schema({vol.Required(CONF_API_TOKEN): str}),
            errors=errors,
        )
```

joostlek commented on homeassistant/components/weatherflow_cloud/coordinator.py:
```diff
@@ -0,0 +1,45 @@
+    def __init__(self, hass: HomeAssistant, api_token: str) -> None:
+        """Initialize global WeatherFlow forecast data updater."""
+
+        # Store local variables
+        self.hass = hass
```
I think this is already set in the update coordinator init

joostlek commented on homeassistant/components/weatherflow_cloud/weather.py:
```diff
@@ -0,0 +1,145 @@
+
+        self.station_id = station_id
+        self._attr_unique_id = f"weatherflow_forecast_{station_id}"
+        self._attr_name = None
```
Can be set outside of the constructor

jeeftor committed:
Update homeassistant/components/weatherflow_cloud/config_flow.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

jeeftor committed:
addressing PR Comments

jeeftor committed:
fixing last comment that i can see

gjohansson-ST commented on homeassistant/components/weatherflow_cloud/const.py:
```diff
@@ -0,0 +1,8 @@
+"""Constants for the WeatherflowCloud integration."""
+import logging
+
+DOMAIN = "weatherflow_cloud"
```
Why isn't this just called `weatherflow`?

jeeftor replied:
Its in the description - but the TLDR is:

Weatherflow provides a UDP based integration completely local and a REST based integration that is cloud based - and uses additional sensors in the area to come up with data

gjohansson-ST replied:
Thanks. Should perhaps have read that first 😁 👍

gjohansson-ST commented on homeassistant/components/weatherflow_cloud/config_flow.py:
```diff
@@ -0,0 +1,50 @@
+                async with WeatherFlowRestAPI(api_token) as api:
+                    await api.async_get_stations()
+            except ClientResponseError as err:
+                if err.status == 401:
+                    errors["base"] = "invalid_api_key"
```
This tells me we should have a `reauth` flow as well and handle that in the coordinator

jeeftor replied:
Reauth Added 👏️

gjohansson-ST commented on homeassistant/components/weatherflow_cloud/coordinator.py:
```diff
@@ -0,0 +1,45 @@
+
+    def __init__(self, hass: HomeAssistant, api_token: str) -> None:
+        """Initialize global WeatherFlow forecast data updater."""
+
+        # Store local variables
+        # self.hass = hass
+
+        # Extract API Token
```
```suggestion
```

gjohansson-ST commented on homeassistant/components/weatherflow_cloud/coordinator.py:
```diff
@@ -0,0 +1,45 @@
+
+    async def _async_update_data(self) -> dict[int, WeatherFlowData]:
+        """Fetch data from WeatherFlow Forecast."""
+        LOGGER.debug("Updating Tempest Data")
```
coordinator is already telling when it's updating
```suggestion
```

gjohansson-ST commented on homeassistant/components/weatherflow_cloud/strings.json:
```diff
@@ -0,0 +1,21 @@
+    "step": {
+      "user": {
+        "description": "Set up a WeatherFlow Forecast Station",
+        "title": "WeatherFlow Forecast",
```
```suggestion
```
Title automatically gets added from the integration name

gjohansson-ST commented on homeassistant/components/weatherflow_cloud/strings.json:
```diff
@@ -0,0 +1,21 @@
+      "unknown": "[%key:common::config_flow::error::unknown%]"
+    },
+    "abort": {
+      "unique_id": "A configuration with this API key has already been configured."
```
`_async_abort_entries_match()` gives back `already_configured`

gjohansson-ST commented on homeassistant/components/weatherflow_cloud/weather.py:
```diff
@@ -0,0 +1,145 @@
+        self._attr_device_info = DeviceInfo(
+            name=self.local_data.station.name,
+            entry_type=DeviceEntryType.SERVICE,
+            identifiers={(DOMAIN, outdoor_device.serial_number)},
```
If this device disappears and a new comes in we suddenly get a stale device and a new device.
Don't understand why it's not `station_id` here?

jeeftor replied:
Ok good point... (also can remove some extra code because I no longer need outdoor device)

gjohansson-ST commented on homeassistant/components/weatherflow_cloud/weather.py:
```diff
@@ -0,0 +1,145 @@
+            identifiers={(DOMAIN, outdoor_device.serial_number)},
+            manufacturer=MANUFACTURER,
+            configuration_url=f"https://tempestwx.com/station/{station_id}/grid",
+            serial_number=outdoor_device.serial_number,
+            sw_version=outdoor_device.firmware_revision,
+            hw_version=outdoor_device.hardware_revision,
```
```suggestion
```

gjohansson-ST commented on homeassistant/components/weatherflow_cloud/weather.py:
```diff
@@ -0,0 +1,145 @@
+    @callback
+    def _async_forecast_daily(self) -> list[Forecast] | None:
+        """Return the daily forecast in native units."""
+        return [x.ha_forecast for x in self.local_data.weather.forecast.daily]
```
As a general remark the library's should not have HA specific things including naming.

gjohansson-ST commented on tests/components/weatherflow_cloud/test_config_flow.py:
```diff
@@ -0,0 +1,44 @@
+from homeassistant.data_entry_flow import FlowResultType
+
+
+async def test_config_2(hass: HomeAssistant, mock_get_stations_combined) -> None:
```
Can we have a bit separate tests so
- Works fine
- Side-effects (better if you parameterize it) which then recover and ends with creating the entry.
- abort because already configured

jeeftor replied:
Done! :)

jeeftor committed:
Update homeassistant/components/weatherflow_cloud/coordinator.py

OOPS

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

jeeftor committed:
Update homeassistant/components/weatherflow_cloud/weather.py

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

jeeftor committed:
Update homeassistant/components/weatherflow_cloud/coordinator.py

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

jeeftor committed:
Update homeassistant/components/weatherflow_cloud/strings.json

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

jeeftor committed:
switching to station id

jeeftor committed:
addressing PR

jeeftor committed:
Updated tests to be better

jeeftor committed:
Updated tests accordingly

jeeftor committed:
REAuth flow and tests added

gjohansson-ST commented on tests/components/weatherflow_cloud/test_config_flow.py:
```diff
@@ -0,0 +1,110 @@
+        )
+        await hass.async_block_till_done()
+
+        assert result["type"] == FlowResultType.FORM
```
Tests should either end with `create_entry` or `abort` so can this test continue by removing the side effect and then successfully finish

gjohansson-ST commented on tests/components/weatherflow_cloud/test_config_flow.py:
```diff
@@ -0,0 +1,110 @@
+        data={CONF_API_TOKEN: "SAME_SAME"},
+    )
+
+    assert result["reason"] == "reauth_successful"
```
Missing result type here.

jeeftor replied:
Uh oh... there is a chance I just pushed to the wrong branch :(

tholland15 commented:
It's probably a long shot, but I would LOVE to see this get added before the forecast property of WeatherEntity gets deprecated in 2024.3

I'm still using the old HACs WeatherFlow (which uses the old forecast property) and have been waiting until this was complete to move to the 2 new core integrations from @jeeftor 

Thanks for all the work you guys do!

jeeftor committed:
Update homeassistant/components/weatherflow_cloud/strings.json

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

jeeftor committed:
Update homeassistant/components/weatherflow_cloud/coordinator.py

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

jeeftor committed:
Addressing PR comments

jeeftor commented:
@tholland15 - there is still a few days left in Feb :) there is always a chance... the local weatherflow got through after the initial release notes were drafted...

gjohansson-ST commented on homeassistant/components/weatherflow_cloud/config_flow.py:
```diff
@@ -0,0 +1,79 @@
+            api_token = user_input[CONF_API_TOKEN]
+            errors = await _validate_api_token(api_token)
+            if not errors:
+                # Update the existing entry and abort
+                existing_entry = self.hass.config_entries.async_get_entry(
+                    self.context["entry_id"]
+                )
+                self.hass.config_entries.async_update_entry(
+                    existing_entry,  # type: ignore[arg-type]
+                    data={CONF_API_TOKEN: api_token},
+                )
+                return self.async_abort(reason="reauth_successful")
```
This is not resolved as the new helper isn't here

gjohansson-ST commented on homeassistant/components/weatherflow_cloud/strings.json:
```diff
@@ -0,0 +1,20 @@
+{
+  "config": {
+    "step": {
+      "user": {
```
unresolved as reauth step still missing

gjohansson-ST commented:
@jeeftor I don't think you pushed any changes before pressing "ready" and also there is conflicts so you need to rebase here to move forward. Thanks
Drafting in the meantime

jeeftor committed:
rebase off dev

jeeftor commented:
@gjohansson-ST - I believe we are in the clear now... you are 100% correct.

I did resolve than push

gjohansson-ST commented on homeassistant/components/weatherflow_cloud/strings.json:
```diff
@@ -0,0 +1,26 @@
+      "reauth": {
+        "description": "Reauthenticate with WeatherFlow",
+        "data": {
+          "api_token": "Personal api token"
```
```suggestion
          "api_token": "[%key:component::weatherflow_cloud::config::step::user::data::api_token%]"
```

gjohansson-ST commented on homeassistant/components/weatherflow_cloud/strings.json:
```diff
@@ -0,0 +1,26 @@
+      "invalid_api_key": "[%key:common::config_flow::error::invalid_api_key%]"
+    },
+
+    "abort": {
+      "already_configured": "A configuration with this API key has already been configured."
+    }
```
```suggestion
    "abort": {
      "already_configured": "[%key:common::config_flow::abort::already_configured_service%]",
      "reauth_successful": "[%key:common::config_flow::abort::reauth_successful%]"
    }
```

gjohansson-ST committed:
Apply suggestions from code review

gjohansson-ST reviewed (approved):
Thanks @jeeftor 👍

gjohansson-ST reviewed (changes requested):
I've marked this PR as draft. CI is failing.
Please un-draft it once it is ready for review again by clicking the "Ready for review" button.

Thanks! 👍

[Learn more about our pull request process.](https://developers.home-assistant.io/docs/review-process#prs-are-being-drafted-when-changes-are-needed)

jeeftor committed:
ruff fix

jeeftor committed:
:x
erge branch 'weatherflow-cloud' of github.com:jeeftor/core into weatherflow-cloud

gjohansson-ST reviewed (approved):
Thanks @jeeftor 👍
